### PR TITLE
[Perf] Fuse LTX-2.3 compile regions for masked FlashAttention cross-attention

### PR DESCRIPTION
**Pull request by mglyn**
# [Perf] Fuse LTX-2.3 compile regions for masked FlashAttention cross-attention

## Purpose

Fuse the compiled LTX-2.3 repeated transformer block into a larger TorchDynamo region when using FlashAttention masked varlen cross-attention.

LTX-2.3 already uses regional compile through the diffusion model runner, but the real pipeline trace showed the compiled block being split around the prompt cross-attention path. The split was not reproduced by a simplified block mock because the real pipeline exercises the masked FlashAttention varlen path.

The graph breaks came from two places:

- FlashAttention backend mask classification branches on `torch.any(~attention_mask)`.
- FlashAttention varlen unpadding uses dynamic shape / scalar ops such as `nonzero()` and `.item()`.

This PR keeps the change local to LTX-2.3:

- Add an LTX2-local `Attention` subclass that can bypass the tensor-valued mask classification branch for marked LTX2 attention layers.
- Add `LTX23Pipeline.setup_compile()` as the pipeline-specific compile entrypoint.
- In LTX-2.3 compile setup, enable Dynamo capture for dynamic output shape ops and scalar outputs before regional compile.
- Mark only LTX-2.3 prompt cross-attention layers (`attn2`, `audio_attn2`) for the local varlen masked path during compile.

The default diffusion attention backend and `AttentionMetadata` are unchanged. Other models keep the existing default compile path.

## Test Plan

Platform:

- 4xA100 80GB

Model:

- LTX-2.3 Diffusers

Workload:

```bash
export PROMPT="Floating crystal islands in cosmic starry sky, glowing nebula, soft luminous particles flowing around, slow camera rotation"
export NEG_PROMPT="low quality, blurry, noise, watermark, text, deformed figures, cartoon style, over-saturated color, frame jump"
```

Compile A/B:

- baseline split: LTX-2.3 compile without dynamic-output capture / local force-varlen path
- fused: LTX-2.3 compile with this PR

Cases:

- `512x384`, 25 frames
- `1024x576`, 81 frames

Each cell:

- warmup with the same shape
- 10 measured requests
- compare `stage_0_gen_ms`

**vLLM Version:**

0.23.0

**vLLM-Omni Commit:**

346bb61eaab1a58fb9faafc9e7d2c6f07ddefd96

## Test Result

Stage-0 generation time:

| attention | exec | graph | case | stage_0_gen_ms_mean | stage_0_gen_ms_p50 | stage_0_gen_ms_p99 | peak_memory_mb_mean | completed | failed |
|---|---|---|---|---:|---:|---:|---:|---:|---:|
| default FLASH_ATTN | compile | fused | 1024x576_81f | 21291.08 | 21276.30 | 21469.63 | 76619.4 | 10 | 0 |
| default FLASH_ATTN | compile | split | 1024x576_81f | 21980.61 | 21978.05 | 22002.86 | 76622.0 | 10 | 0 |
| default FLASH_ATTN | compile | fused | 512x384_25f | 3914.43 | 3888.54 | 4114.49 | 73259.0 | 10 | 0 |
| default FLASH_ATTN | compile | split | 512x384_25f | 4841.15 | 4815.68 | 4998.12 | 73259.4 | 10 | 0 |

Observed improvement:

| case | mean delta | relative |
|---|---:|---:|
| 1024x576_81f | -689.54 ms | -3.1% |
| 512x384_25f | -926.72 ms | -19.1% |

Peak memory was effectively unchanged.
---
Trace comparison:

8f0a6e27-e781-49bc-9674-be552efdb61c
- Before: the block is split into several Torch-Compiled Regions around the prompt cross-attention path.

---

91b42093-0e24-4074-832c-f05969bcf735
- After: the same block is captured as a larger continuous Torch-Compiled Region, reducing the CPU-side gaps between compiled fragments.

---

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py [hunk 1]
```diff
@@ -75,6 +75,27 @@ def _get_audio_latents_from_sampling(sampling: Any) -> torch.Tensor | None:
     LTX2VocoderWithBWE = None
 
 
+def _setup_ltx23_regional_compile(pipeline: nn.Module) -> None:
+    # LTX-2.3 prompt cross-attention uses FlashAttention varlen unpadding.
+    # Capturing dynamic output shape ops keeps the repeated transformer block
+    # in one compiled region instead of splitting around nonzero()/item().
+    torch._dynamo.config.capture_dynamic_output_shape_ops = True
```
Comment (david6666666) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py:82]: These Dynamo flags are process-wide, so this leaks outside LTX-2.3 after setup_compile() runs. Can we scope them with torch._dynamo.config.patch(...) around the regional compile?

Comment (mglyn) [reply vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py:82]: Thanks, agreed. I ended up moving away from this approach.

I found a smaller LTX-2.3-local implementation in #4797 that avoids touching `setup_compile()` entirely, so it does not need any process-wide Dynamo config changes.

The new approach is based on the LTX-2.3 learned-register connector behavior: when both video/audio connectors use learned registers, padding positions are already replaced by learned register embeddings and the connector returns an all-valid downstream mask. In that case, passing `None` to the denoise transformer is equivalent to passing the all-valid mask, while avoiding the masked FlashAttention varlen path and its graph-break / bookkeeping overhead.

So #4797:
- does not modify `torch._dynamo.config`
- does not override `setup_compile()`
- does not touch shared attention backends
- keeps the old mask path when connectors do not expose learned registers

I also reran a more complete A100 sweep with profiler disabled, same-shape warmup, 10 measured requests per cell, covering base/cfg2 and 512x384x25f / 1024x576x81f. The result shows consistent `stage_0_gen_ms` improvement with effectively unchanged peak memory.

If #4797 looks good, I will close this PR in favor of that smaller implementation.

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py [hunk 2]
```diff
@@ -75,6 +75,27 @@ def _get_audio_latents_from_sampling(sampling: Any) -> torch.Tensor | None:
     LTX2VocoderWithBWE = None
```
Comment (wtomin) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py:75]: Could you update `tests/diffusion/models/ltx2/test_ltx2_3_pipeline.py` to add regression test for this feature?

File: vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py (hunk 1 above)
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2/pipeline_ltx2_3.py:82]: These two flip global `torch._dynamo` config process-wide and never restore it, so every model compiled after this in the same process inherits `capture_dynamic_output_shape_ops`/`capture_scalar_outputs` and gets its graph-break behavior changed underneath it. They're also set before the `transformer is None` guard, so a pipeline with no transformer still pays for it. Can we save/restore the prior values and move them below the guard? 🤔

File: vllm_omni/diffusion/models/ltx2/ltx2_transformer.py [hunk 3]
```diff
@@ -516,7 +539,12 @@ def __call__(
         key = key.unflatten(2, (attn.heads, attn.head_dim))
         value = value.unflatten(2, (attn.heads, attn.head_dim))
 
-        attn_metadata = AttentionMetadata(attn_mask=attention_mask) if attention_mask is not None else None
+        attn_metadata = None
+        if attention_mask is not None:
+            extra = {}
+            if attn.force_flash_varlen_mask and torch.compiler.is_compiling():
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2/ltx2_transformer.py:545]: The extra key only gets set under `torch.compiler.is_compiling()`, so the varlen masked path is unreachable in eager -- anyone who flips `force_flash_varlen_mask` expecting the FlashAttention fast path outside compile silently falls back to the dense mask. If that's intentional (reads like it is, given the region-fusion goal), a one-line comment right here saves the next person the head-scratch 🙏

File: vllm_omni/diffusion/models/ltx2/ltx2_transformer.py [hunk 4]
```diff
@@ -564,6 +592,7 @@ def __init__(
         quant_config: "QuantizationConfig | None" = None,
         prefix: str = "",
         disable_kv_quant: bool = False,
+        force_flash_varlen_mask: bool = False,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/ltx2/ltx2_transformer.py:595]: `force_flash_varlen_mask` is threaded through the constructor but nothing ever passes it -- attn2/audio_attn2 are built with the default, and the only place it gets turned on is `_setup_ltx23_regional_compile` mutating the attribute after the fact. Either drop the ctor arg and keep that single mutation point, or set it at construction; carrying both mechanisms is confusing.

**Comment by mglyn** [comment]
This issue doesn't exist anymore after #5148 was merged.
